File: desktop/main/protocol.js
const TRACKS = new Set(["system", "microphone"]);
const ENGINE_STATUSES = new Set([
  "configured",
  "loading",
  "ready",
  "flushed",
  "shutdown",
  "unavailable"
]);
const MODEL_LOAD_PHASES = new Set([
  "checking_cache",
  "downloading",
  "verifying",
  "initializing",
  "preparing_speakers",
  "checking_translation_cache",
  "downloading_translation",
  "verifying_translation",
  "converting_translation",
  "initializing_translation"
]);
const EVENT_TYPES = new Set([
  "engine_status",
  "model_progress",
  "warning",
  "error",
  "partial_transcript",
  "final_segment",
  "session_stopped"
]);

export const MAX_AUDIO_PACKET_BYTES = 64_000;
export const MAX_BACKEND_LINE_BYTES = 2_000_000;

export class ProtocolValidationError extends Error {
  constructor(message, code = "invalid_protocol") {
    super(message);
    this.name = "ProtocolValidationError";
    this.code = code;
  }
}

export function parseBackendLine(line) {
  if (typeof line !== "string" || line.length === 0) {
    throw new ProtocolValidationError("Backend output must be a non-empty JSON line.");
  }

  let value;
  try {
    value = JSON.parse(line);
  } catch {
    throw new ProtocolValidationError("Backend emitted malformed JSON.", "malformed_backend_output");
  }

  return validateBackendEvent(value);
}

export function validateBackendEvent(value) {
  const event = requireRecord(value, "Backend event");
  const type = requireString(event.type, "event.type", 64);
  if (!EVENT_TYPES.has(type)) {
    throw new ProtocolValidationError(`Unsupported backend event type: ${type}`);
  }

  if (type === "engine_status") {
    const status = requireString(event.status, "engine_status.status", 64);
    if (!ENGINE_STATUSES.has(status)) {
      throw new ProtocolValidationError(`Unsupported engine status: ${status}`);
    }
    return compactObject({
      type,
      status,
      session_id: optionalString(event.session_id, "engine_status.session_id", 256),
      language: optionalString(event.language, "engine_status.language", 64),
      device: optionalString(event.device, "engine_status.device", 64),
      compute: optionalString(event.compute, "engine_status.compute", 64)
    });
  }

  if (type === "model_progress") {
    const phase = requireString(event.phase, "model_progress.phase", 64);
    if (!MODEL_LOAD_PHASES.has(phase)) {
      throw new ProtocolValidationError(`Unsupported model load phase: ${phase}`);
    }
    return compactObject({
      type,
      phase,
      session_id: optionalString(event.session_id, "model_progress.session_id", 256)
    });
  }

  if (type === "warning" || type === "error") {
    return compactObject({
      type,
      source: requireOneOf(event.source, ["protocol", "capture", "transcription"], "issue.source"),
      code: requireString(event.code, "issue.code", 128),
      message: requireString(event.message, "issue.message", 2_000),
      recoverable: requireBoolean(event.recoverable, "issue.recoverable"),
      segment_id: optionalString(event.segment_id, "issue.segment_id", 256)
    });
  }

  if (type === "partial_transcript" || type === "final_segment") {
    return {
      type,
      session_id: requireString(event.session_id, "transcript.session_id", 256),
      segment: validateSegment(event.segment, type)
    };
  }

  return compactObject({
    type,
    session_id: optionalString(event.session_id, "session_stopped.session_id", 256),
    reason: optionalString(event.reason, "session_stopped.reason", 256)
  });
}

export function createStartCommand(options = {}) {
  const input = requireRecord(options, "Start options");
  const command = { type: "start" };
  for (const key of [
    "model", "language", "device", "compute", "download_root", "diarization_model", "translation_model"
  ]) {
    if (!(key in input) || input[key] === undefined) continue;
    if ((key === "language" || key === "download_root" || key === "diarization_model" || key === "translation_model") && input[key] === null) {
      command[key] = null;
      continue;
    }
    command[key] = requireString(input[key], `start.${key}`, 1_024);
  }
  if (input.diarization !== undefined) {
    command.diarization = requireOneOf(input.diarization, ["off", "online"], "start.diarization");
  }
  if (input.translation !== undefined) {
    command.translation = requireOneOf(input.translation, ["off", "en_to_pt_br"], "start.translation");
  }
  return command;
}

export function createAudioCommand(packet) {
  const input = requireRecord(packet, "Audio packet");
  const track = requireOneOf(input.track, [...TRACKS], "audio.track");
  const startMs = requireInteger(input.startMs, "audio.startMs", 0);
  const endMs = requireInteger(input.endMs, "audio.endMs", 1);
  if (endMs <= startMs) {
    throw new ProtocolValidationError("audio.endMs must be greater than audio.startMs.");
  }

  const pcm = toUint8Array(input.pcm);
  if (pcm.byteLength === 0 || pcm.byteLength % 2 !== 0) {
    throw new ProtocolValidationError("Audio PCM must contain complete signed 16-bit samples.");
  }
  if (pcm.byteLength > MAX_AUDIO_PACKET_BYTES) {
    throw new ProtocolValidationError("Audio packet exceeds the desktop safety limit.");
  }
  const durationMs = Math.max(1, Math.round((pcm.byteLength / 2) * 1_000 / 16_000));
  if (endMs !== startMs + durationMs) {
    throw new ProtocolValidationError("Audio timestamps do not match the 16 kHz PCM sample count.");
  }

  return {
    type: "audio",
    track,
    start_ms: startMs,
    end_ms: endMs,
    pcm_s16le_base64: Buffer.from(pcm.buffer, pcm.byteOffset, pcm.byteLength).toString("base64")
  };
}

function validateSegment(value, eventType) {
  const segment = requireRecord(value, "Transcript segment");
  const partial = requireBoolean(segment.partial, "segment.partial");
  const final = requireBoolean(segment.final, "segment.final");
  if (partial === final) {
    throw new ProtocolValidationError("Segment partial and final flags must be complementary.");
  }
  if (eventType === "partial_transcript" && !partial) {
    throw new ProtocolValidationError("Partial event requires segment.partial=true.");
  }
  if (eventType === "final_segment" && !final) {
    throw new ProtocolValidationError("Final event requires segment.final=true.");
  }

  const startMs = requireInteger(segment.start_ms, "segment.start_ms", 0);
  const endMs = requireInteger(segment.end_ms, "segment.end_ms", 0);
  if (endMs < startMs) {
    throw new ProtocolValidationError("segment.end_ms must not be before segment.start_ms.");
  }

  const translatedText = optionalString(segment.translated_text, "segment.translated_text", 20_000);
  const translatedLanguage = optionalString(segment.translated_language, "segment.translated_language", 16);
  if (translatedText === undefined) {
    if (translatedLanguage !== undefined) {
      throw new ProtocolValidationError("segment.translated_language must be null when translated text is absent.");
    }
  } else {
    if (!final) throw new ProtocolValidationError("Partial segments cannot contain translated text.");
    if (translatedLanguage !== "pt-BR") {
      throw new ProtocolValidationError("Translated segments require segment.translated_language=pt-BR.");
    }
  }

  return compactObject({
    id: requireString(segment.id, "segment.id", 256),
    revision: requireInteger(segment.revision, "segment.revision", 0),
    start_ms: startMs,
    end_ms: endMs,
    track: requireOneOf(segment.track, [...TRACKS], "segment.track"),
    text: requireString(segment.text, "segment.text", 20_000, true),
    partial,
    final,
    language: optionalString(segment.language, "segment.language", 64),
    speaker_id: optionalString(segment.speaker_id, "segment.speaker_id", 256),
    translated_text: translatedText,
    translated_language: translatedLanguage
  });
}

function requireRecord(value, label) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new ProtocolValidationError(`${label} must be an object.`);
  }
  return value;
}

function requireString(value, label, maxLength, allowEmpty = false) {
  if (typeof value !== "string" || (!allowEmpty && value.trim().length === 0) || value.length > maxLength) {
    throw new ProtocolValidationError(`${label} must be a valid string.`);
  }
  return value;
}

function optionalString(value, label, maxLength) {
  if (value === undefined || value === null) return undefined;
  return requireString(value, label, maxLength, true);
}

function requireBoolean(value, label) {
  if (typeof value !== "boolean") {
    throw new ProtocolValidationError(`${label} must be a boolean.`);
  }
  return value;
}

function requireInteger(value, label, minimum) {
  if (!Number.isSafeInteger(value) || value < minimum) {
    throw new ProtocolValidationError(`${label} must be an integer of at least ${minimum}.`);
  }
  return value;
}

function requireOneOf(value, allowed, label) {
  if (!allowed.includes(value)) {
    throw new ProtocolValidationError(`${label} has an unsupported value.`);
  }
  return value;
}

function toUint8Array(value) {
  if (value instanceof Uint8Array) return value;
  if (value instanceof ArrayBuffer) return new Uint8Array(value);
  if (ArrayBuffer.isView(value)) return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
  throw new ProtocolValidationError("audio.pcm must be an ArrayBuffer or typed array.");
}

function compactObject(value) {
  return Object.fromEntries(Object.entries(value).filter(([, entry]) => entry !== undefined));
}

File: backend/src/meeting_transcriber/protocol.py
"""Typed JSON Lines command parsing and event serialization."""

from __future__ import annotations

from dataclasses import asdict, dataclass
import json
from typing import Any, Literal, Mapping, TypeAlias

from .audio import PcmValidationError, decode_pcm_s16le_base64, duration_ms


Track: TypeAlias = Literal["system", "microphone"]
MAX_SEGMENT_TEXT_CHARS = 20_000


class ProtocolError(ValueError):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True, slots=True)
class ConfigureCommand:
    changes: dict[str, object]


@dataclass(frozen=True, slots=True)
class StartCommand:
    changes: dict[str, object]


@dataclass(frozen=True, slots=True)
class AudioCommand:
    track: Track
    start_ms: int
    end_ms: int
    pcm_s16le: bytes


@dataclass(frozen=True, slots=True)
class StopCommand:
    pass


@dataclass(frozen=True, slots=True)
class FlushCommand:
    pass


@dataclass(frozen=True, slots=True)
class ShutdownCommand:
    pass


Command: TypeAlias = (
    ConfigureCommand | StartCommand | AudioCommand | StopCommand | FlushCommand | ShutdownCommand
)


@dataclass(frozen=True, slots=True)
class SegmentPayload:
    id: str
    revision: int
    start_ms: int
    end_ms: int
    track: Track
    text: str
    partial: bool
    final: bool
    language: str | None
    speaker_id: str | None = None
    translated_text: str | None = None
    translated_language: str | None = None

    def __post_init__(self) -> None:
        _validate_text_field(self.text, "Segment text", nullable=False)
        _validate_text_field(self.translated_text, "Translated segment text", nullable=True)
        if self.partial and self.translated_text is not None:
            raise ValueError("Partial segments must not contain translated text")
        if self.translated_text is None:
            if self.translated_language is not None:
                raise ValueError("translated_language must be null when translated_text is null")
        elif self.translated_language != "pt-BR":
            raise ValueError("translated_language must be 'pt-BR' when translated_text is present")


_SETTING_FIELDS = {
    "model",
    "language",
    "device",
    "compute",
    "download_root",
    "diarization",
    "diarization_model",
    "translation",
    "translation_model",
}


def parse_command(line: str) -> Command:
    try:
        payload = json.loads(line)
    except json.JSONDecodeError as exc:
        raise ProtocolError("invalid_json", f"Invalid JSON at character {exc.pos}") from exc
    if not isinstance(payload, dict):
        raise ProtocolError("invalid_command", "Each line must be a JSON object")

    command_type = payload.get("type")
    if not isinstance(command_type, str):
        raise ProtocolError("invalid_command", "Command field 'type' must be a string")

    if command_type in {"configure", "start"}:
        _reject_unknown(payload, {"type", *_SETTING_FIELDS})
        changes = _parse_settings(payload)
        return ConfigureCommand(changes) if command_type == "configure" else StartCommand(changes)
    if command_type == "audio":
        _reject_unknown(payload, {"type", "track", "start_ms", "end_ms", "pcm_s16le_base64"})
        return _parse_audio(payload)
    if command_type in {"stop", "flush", "shutdown"}:
        _reject_unknown(payload, {"type"})
        return {
            "stop": StopCommand,
            "flush": FlushCommand,
            "shutdown": ShutdownCommand,
        }[command_type]()
    raise ProtocolError("unknown_command", f"Unknown command type '{command_type}'")


def _parse_settings(payload: Mapping[str, Any]) -> dict[str, object]:
    changes: dict[str, object] = {}
    for field in _SETTING_FIELDS:
        if field not in payload:
            continue
        value = payload[field]
        if field in {"language", "download_root", "diarization_model", "translation_model"} and value is None:
            changes[field] = None
        elif field == "diarization":
            if not isinstance(value, str) or value not in {"off", "online"}:
                raise ProtocolError("invalid_configuration", "'diarization' must be 'off' or 'online'")
            changes[field] = value
        elif field == "translation":
            if not isinstance(value, str) or value not in {"off", "en_to_pt_br"}:
                raise ProtocolError(
                    "invalid_configuration",
                    "'translation' must be 'off' or 'en_to_pt_br'",
                )
            changes[field] = value
        elif not isinstance(value, str) or not value.strip():
            raise ProtocolError("invalid_configuration", f"'{field}' must be a non-empty string or null")
        else:
            normalized = value.strip()
            if field in {"diarization_model", "translation_model"} and (
                len(normalized) > 4_096 or "\x00" in normalized
            ):
                raise ProtocolError(
                    "invalid_configuration",
                    f"'{field}' must be a valid local path or null",
                )
            changes[field] = normalized
    return changes


def _parse_audio(payload: Mapping[str, Any]) -> AudioCommand:
    track = payload.get("track")
    if track not in {"system", "microphone"}:
        raise ProtocolError("invalid_audio", "track must be 'system' or 'microphone'")
    start_ms = payload.get("start_ms")
    if isinstance(start_ms, bool) or not isinstance(start_ms, int) or start_ms < 0:
        raise ProtocolError("invalid_audio", "start_ms must be a non-negative integer")
    try:
        pcm = decode_pcm_s16le_base64(payload.get("pcm_s16le_base64"))
    except PcmValidationError as exc:
        raise ProtocolError("invalid_audio", str(exc)) from exc
    calculated_end_ms = start_ms + duration_ms(pcm)
    supplied_end_ms = payload.get("end_ms", calculated_end_ms)
    if isinstance(supplied_end_ms, bool) or not isinstance(supplied_end_ms, int):
        raise ProtocolError("invalid_audio", "end_ms must be an integer when supplied")
    if supplied_end_ms != calculated_end_ms:
        raise ProtocolError(
            "invalid_audio",
            "end_ms does not match the duration of the 16 kHz PCM packet",
        )
    return AudioCommand(track=track, start_ms=start_ms, end_ms=calculated_end_ms, pcm_s16le=pcm)


def _reject_unknown(payload: Mapping[str, Any], allowed: set[str]) -> None:
    unknown = sorted(set(payload) - allowed)
    if unknown:
        raise ProtocolError("invalid_command", f"Unknown field(s): {', '.join(unknown)}")


def serialize_event(event: Mapping[str, Any]) -> str:
    event_type = event.get("type")
    if not isinstance(event_type, str) or not event_type:
        raise ProtocolError("invalid_event", "Event must have a non-empty type")
    if event_type in {"partial_transcript", "final_segment"}:
        _validate_segment_event(event_type, event)
    return json.dumps(dict(event), ensure_ascii=False, separators=(",", ":"))


def segment_event(
    event_type: Literal["partial_transcript", "final_segment"],
    session_id: str,
    segment: SegmentPayload,
) -> dict[str, object]:
    payload = asdict(segment)
    event: dict[str, object] = {"type": event_type, "session_id": session_id, "segment": payload}
    _validate_segment_event(event_type, event)
    return event


def _validate_segment_event(event_type: str, event: Mapping[str, Any]) -> None:
    session_id = event.get("session_id")
    if not isinstance(session_id, str) or not session_id:
        raise ProtocolError("invalid_event", "Transcript event requires a non-empty session_id")
    segment = event.get("segment")
    if not isinstance(segment, Mapping):
        raise ProtocolError("invalid_event", "Transcript event requires a segment object")
    partial = segment.get("partial")
    final = segment.get("final")
    if not isinstance(partial, bool) or not isinstance(final, bool) or partial == final:
        raise ProtocolError("invalid_event", "Segment partial and final flags must be complementary")
    if event_type == "partial_transcript" and not partial:
        raise ProtocolError("invalid_event", "Partial transcript event must have partial=true")
    if event_type == "final_segment" and not final:
        raise ProtocolError("invalid_event", "Final segment event must have final=true")
    try:
        _validate_text_field(segment.get("text"), "Segment text", nullable=False)
        translated_text = segment.get("translated_text")
        translated_language = segment.get("translated_language")
        _validate_text_field(translated_text, "Translated segment text", nullable=True)
        if partial and translated_text is not None:
            raise ValueError("Partial segments must not contain translated text")
        if translated_text is None:
            if translated_language is not None:
                raise ValueError("translated_language must be null when translated_text is null")
        elif translated_language != "pt-BR":
            raise ValueError("translated_language must be 'pt-BR' when translated_text is present")
    except ValueError as exc:
        raise ProtocolError("invalid_event", str(exc)) from exc


def _validate_text_field(value: object, label: str, *, nullable: bool) -> None:
    if value is None and nullable:
        return
    if not isinstance(value, str) or len(value) > MAX_SEGMENT_TEXT_CHARS or "\x00" in value:
        qualifier = "a bounded string or null" if nullable else "a bounded string"
        raise ValueError(f"{label} must be {qualifier}")


def issue_event(
    kind: Literal["warning", "error"],
    *,
    source: Literal["protocol", "capture", "transcription"],
    code: str,
    message: str,
    recoverable: bool,
    segment_id: str | None = None,
) -> dict[str, object]:
    event: dict[str, object] = {
        "type": kind,
        "source": source,
        "code": code,
        "message": message,
        "recoverable": recoverable,
    }
    if segment_id is not None:
        event["segment_id"] = segment_id
    return event
